server.go

```diff
@@ -237,3 +240,58 @@ type ScanCloser struct {
 	*bufio.Scanner
 	closer TimeoutCloser
 }
+
+type DatagramMessage struct {
+	message []byte
+	client  string
+}
+
+func (s *Server) goReceiveDatagrams(connection net.Conn) {
+	packetconn, ok := connection.(net.PacketConn)
+	if !ok {
+		panic("Connection is not a packet connection")
+	}
+	s.wait.Add(1)
+	go func() {
+		defer s.wait.Done()
+		for {
+			buf := make([]byte, 65536)
+			n, addr, err := packetconn.ReadFrom(buf)
+			if err == nil {
+				// Ignore trailing control characters and NULs
+				for ; (n > 0) && (buf[n-1] < 32); n-- {
```

Comment:
From RFC3164 section 4.1.3:

> The MSG part of the syslog packet MUST contain visible (printing)
> characters. The code set traditionally and most often used has also
> been seven-bit ASCII in an eight-bit field like that used in the PRI
> and HEADER parts.  In this code set, the only allowable characters
> are the ABNF VCHAR values (%d33-126) and spaces (SP value %d32).

I am reading the first sentence to mean "the MSG part of the syslog
packet MUST contain *only* visible (printing) characters". I think this
is reasonably uncontroversial as otherwise the following two sentences
would not make sense.

What the code supplied does is eliminate trailing control characters
(less than code 32) that per the RFC should not be there anyway. As
they are supplied in contravention of the RFC there should be no objection
to deleting them, and they may cause a parser (or something displaying
the syslog) difficulty. The prime offenders are going to be \0, \n, \r
and (possibly) \t. Given including these is contrary to the specification,
is is most likely they are inserted erroneously. But rather than
pontificating as to whether (e.g.) vertical tab should be stripped, I
thought the best option was simply to remove them, as they clearly
should not be there anyway.

That's RFC3164 (majority of syslog), but this is code path is
also used by RFC5424, so let's have a look at that (at 6.4):

> The syslog application SHOULD avoid octet values below 32 (the
> traditional US-ASCII control character range except DEL).  These
> values are legal, but a syslog application MAY modify these
> characters upon reception.  For example, it might change them into an
> escape sequence (e.g., value 0 may be changed to "\0").  A syslog
> application SHOULD NOT modify any other octet values.

This is a weaker statement, and says we should not have been sent
anything below 32, but if we are, we are entitled to change them. One
possibility is that we escape them (as suggested) but another possibility
is that we simply delete them (if they are trailing). What I'm proposing
falls within that 'MAY'.

Being less than 32, these characters can never form part of a multibyte
unicode code point.

I'd be happy with something that stripped \0, \n and \r but the code
would be longer and slower to handle just those cases. IE I think the
code as presented is correct.


   If a syslog application encodes MSG in UTF-8, the string MUST sta

On 5 Jul 2015, at 21:08, M�ximo Cuadros <notifications@github.com> wrote:

> In server.go:
> 
> > +}
> > +
> > +func (s *Server) goReceiveDatagrams(connection net.Conn) {
> > +	packetconn, ok := connection.(net.PacketConn)
> > +	if !ok {
> > +		panic("Connection is not a packet connection")
> > +	}
> > +	s.wait.Add(1)
> > +	go func() {
> > +		defer s.wait.Done()
> > +		for {
> > +			buf := make([]byte, 65536)
> > +			n, addr, err := packetconn.ReadFrom(buf)
> > +			if err == nil {
> > +				// Ignore trailing control characters and NULs
> > +				for ; (n > 0) && (buf[n-1] < 32); n-- {
> 
> This ignore more than nul and \n why?
> 
> �
> Reply to this email directly or view it on GitHub.
> 

-- 
Alex Bligh